go/types/subtype_test.go

```diff
@@ -437,3 +437,206 @@ func TestIsSubtypeOptionalFields(tt *testing.T) {
+	assertTrue(NewMap(), MakeMapType(MakeUnionType(), MakeUnionType()))
+
+	{
+		newChunkedMap := func(vs ...Value) Map {
+			newSequenceMetaTuple := func(e mapEntry) metaTuple {
+				seq := newMapLeafSequence(nil, e)
+				m := newMap(seq)
+				return newMetaTuple(NewRef(m), newOrderedKey(e.key), 1, m)
+			}
+
+			tuples := make([]metaTuple, len(vs)/2)
+			for i := 0; i < len(vs); i += 2 {
+				tuples[i/2] = newSequenceMetaTuple(mapEntry{vs[i], vs[i+1]})
+			}
+			return newMap(newMapMetaSequence(tuples, nil))
+		}
+
+		assertTrue(newChunkedMap(Number(0), String("a"), Number(1), String("b")), MakeMapType(NumberType, StringType))
+		assertFalse(newChunkedMap(Number(0), String("a"), Number(1), String("b")), MakeMapType(BoolType, StringType))
+		assertFalse(newChunkedMap(Number(0), String("a"), Number(1), String("b")), MakeMapType(NumberType, BoolType))
+		assertTrue(newChunkedMap(Number(0), String("a"), Number(1), String("b")), MakeMapType(MakeUnionType(NumberType, BoolType), StringType))
+		assertTrue(newChunkedMap(Number(0), String("a"), Number(1), String("b")), MakeMapType(NumberType, MakeUnionType(BoolType, StringType)))
+		assertTrue(newChunkedMap(Number(0), String("a"), Bool(true), String("b")), MakeMapType(MakeUnionType(NumberType, BoolType), StringType))
+		assertTrue(newChunkedMap(Number(0), String("a"), Number(1), Bool(true)), MakeMapType(NumberType, MakeUnionType(BoolType, StringType)))
+		assertFalse(newChunkedMap(Number(0), String("a")), MakeMapType(MakeUnionType(), StringType))
+		assertFalse(newChunkedMap(Number(0), String("a")), MakeMapType(NumberType, MakeUnionType()))
+	}
+
+	assertTrue(NewRef(Number(1)), MakeRefType(NumberType))
+	assertFalse(NewRef(Number(1)), MakeRefType(BoolType))
+	assertTrue(NewRef(Number(1)), MakeRefType(MakeUnionType(NumberType, BoolType)))
+	assertFalse(NewRef(Number(1)), MakeRefType(MakeUnionType()))
+
+	assertTrue(
+		NewStruct("Struct", StructData{"x": Bool(true)}),
+		MakeStructType("Struct", StructField{"x", BoolType, false}),
+	)
+	assertTrue(
+		NewStruct("Struct", StructData{"x": Bool(true)}),
+		MakeStructType("Struct", StructField{"x", BoolType, true}),
+	)
+	assertTrue(
+		NewStruct("Struct", StructData{"x": Bool(true)}),
+		MakeStructType("Struct"),
+	)
+	assertTrue(
+		NewStruct("Struct", StructData{}),
+		MakeStructType("Struct"),
+	)
+	assertFalse(
+		NewStruct("", StructData{"x": Bool(true)}),
+		MakeStructType("Struct"),
+	)
+	assertFalse(
+		NewStruct("struct", StructData{"x": Bool(true)}), // lower case name
+		MakeStructType("Struct"),
+	)
+	assertTrue(
+		NewStruct("Struct", StructData{"x": Bool(true)}),
+		MakeStructType("Struct", StructField{"x", MakeUnionType(BoolType, NumberType), true}),
+	)
+	assertTrue(
+		NewStruct("Struct", StructData{"x": Bool(true)}),
+		MakeStructType("Struct", StructField{"y", BoolType, true}),
+	)
+	assertFalse(
+		NewStruct("Struct", StructData{"x": Bool(true)}),
+		MakeStructType("Struct", StructField{"x", StringType, true}),
+	)
+
+	assertTrue(
```

Comment:
a bit worried the cycle case isn't detected correctly. what about the following case:

```
Type = Struct Node {
  value: Number
  children: List<Cycle<Node>>
}

Value = Struct Node {
  value: 1
  children: List(Struct Node {
    value: false
    children: List()
  }
}